src/org/sosy_lab/java_smt/test/SolverContextTest.java

```diff
@@ -119,4 +122,66 @@ public void testFormulaAccessAfterClose() {
     assertThat(bmgr.isTrue(opTerm)).isFalse();
     assertThat(bmgr.isFalse(opTerm)).isFalse();
   }
+
+  @Test
+  public void testProverCopying() throws SolverException, InterruptedException {
+    requireProverCopying();
+    try (ProverEnvironment prover = context.newProverEnvironment()) {
+      assertThat(prover.isUnsat()).isFalse();
+      try (ProverEnvironment copiedProver = context.copyProverEnvironment(prover)) {
+        assertThat(copiedProver.isUnsat()).isFalse();
+      }
+    }
+  }
+
+  @SuppressWarnings("resource")
+  @Test
+  public void testProverCopyCloseInitialProver() throws SolverException, InterruptedException {
+    requireProverCopying();
+    ProverEnvironment prover = context.newProverEnvironment();
+    try (ProverEnvironment copiedProver = context.copyProverEnvironment(prover)) {
+      prover.close();
+      assertThat(copiedProver.isUnsat()).isFalse();
+    }
+  }
+
+  /*
+   * Create a prover, push a SAT and an UNSAT formula on 2 levels, copy the prover, check that
+   * the stack is copied correctly.
+   */
+  @Test
+  public void testProverCopyWithStackAndAssertions() throws InterruptedException, SolverException {
+    requireProverCopying();
+
+    assume()
+        .withMessage(
+            "Solver %s does not support prover copying for non-base-level provers",
+            solverToUse())
+        .that(solverToUse())
+        .isNotEqualTo(Solvers.Z3);
```

Comment:
I do not understand this test: Is the whole copying-thing working or not?

Reply:
Yes. For a stack that has not been pushed according to the error message. Which makes it unusable for the use-case of Thomas. I've updated the discussion in the Z3 repo to find out if i made a mistake or if this is intended behavior.